"error: use of class template '::b' requires template arguments"

Enable the test case `test_issue_470`, which should have been fixed by #481. We now encounter this instead.

```
cargo:warning=/var/folders/qm/fz111zfj06vd69gclc3pbcf400m02d/T/.tmpSBR36v/target/cxx/gen0.cxx:125:45: error: use of class template '::b' requires template arguments
cargo:warning=static_assert(::rust::detail::is_complete<::b>::value, "definition of b is required");
cargo:warning=                                            ^
cargo:warning=/var/folders/qm/fz111zfj06vd69gclc3pbcf400m02d/T/.tmpSBR36v/input.h:5:28: note: template is declared here
cargo:warning=template <typename> struct b { template <typename> struct c; };
cargo:warning=~~~~~~~~~~~~~~~~~~~        ^

```

--- a/engine/Cargo.toml
+++ b/engine/Cargo.toml
@@ -31,7 +31,7 @@ log = "0.4"
 proc-macro2 = "1.0.11"
 quote = "1.0"
 indoc = "1.0"
-autocxx-bindgen = "0.58.4"
+autocxx-bindgen = "0.58.5"
 itertools = "0.9"
 cc = { version = "1.0", optional = true }
 unzip-n = "0.1.2"

--- a/engine/src/conversion/analysis/fun/mod.rs
+++ b/engine/src/conversion/analysis/fun/mod.rs
@@ -19,7 +19,10 @@ mod rust_name_tracker;
 
 use crate::{
     conversion::{
-        analysis::type_converter::{add_analysis, TypeConversionContext, TypeConverter},
+        analysis::{
+            has_attr,
+            type_converter::{add_analysis, TypeConversionContext, TypeConverter},
+        },
         api::TypedefKind,
         convert_error::ConvertErrorWithContext,
         convert_error::ErrorContext,
@@ -413,7 +416,7 @@ impl<'a> FnAnalyzer<'a> {
             } else if is_static_method {
                 MethodKind::Static
             } else if param_details.iter().any(|pd| pd.is_virtual) {
-                if Self::has_attr(&fun, "bindgen_pure_virtual") {
+                if has_attr(&fun.attrs, "bindgen_pure_virtual") {
                     MethodKind::PureVirtual
                 } else {
                     MethodKind::Virtual
@@ -462,7 +465,7 @@ impl<'a> FnAnalyzer<'a> {
             }
         }
         // Second, reject any functions handling types which we flake out on.
-        if Self::has_attr(&fun, "bindgen_unused_template_param_in_arg_or_return") {
+        if has_attr(&fun.attrs, "bindgen_unused_template_param_in_arg_or_return") {
             return Err(contextualize_error(ConvertError::UnusedTemplateParam));
         }
 
@@ -827,10 +830,6 @@ impl<'a> FnAnalyzer<'a> {
             }
         }
         (ref_params, ref_return)
-    }
-
-    fn has_attr(fun: &ForeignItemFn, attr_name: &str) -> bool {
-        fun.attrs.iter().any(|at| at.path.is_ident(attr_name))
     }
 }
 

--- a/engine/src/conversion/analysis/fun/mod.rs
+++ b/engine/src/conversion/analysis/fun/mod.rs
@@ -26,7 +26,7 @@ use crate::{
         api::TypedefKind,
         convert_error::ConvertErrorWithContext,
         convert_error::ErrorContext,
-        error_reporter::add_api_or_report_error,
+        error_reporter::convert_apis,
     },
     known_types::known_types,
 };
@@ -152,9 +152,7 @@ impl<'a> FnAnalyzer<'a> {
             generate_utilities: Self::should_generate_utilities(&apis),
         };
         let mut results = Vec::new();
-        for api in apis {
-            add_api_or_report_error(api.name(), &mut results, || me.analyze_fn_api(api));
-        }
+        convert_apis(apis, &mut results, |api| me.analyze_fn_api(api));
         results.extend(me.extra_apis.into_iter().map(add_analysis));
         results
     }

--- a/engine/src/conversion/analysis/mod.rs
+++ b/engine/src/conversion/analysis/mod.rs
@@ -14,6 +14,8 @@
 
 use syn::Attribute;
 
+use super::ConvertError;
+
 pub(crate) mod abstract_types;
 pub(crate) mod ctypes;
 pub(crate) mod fun;
@@ -25,7 +27,13 @@ mod type_converter;
 
 // Remove `bindgen_` attributes. They don't have a corresponding macro defined anywhere,
 // so they will cause compilation errors if we leave them in.
-fn remove_bindgen_attrs(attrs: &mut Vec<Attribute>) {
+// We may return an error if one of the bindgen attributes shows that the
+// item can't be processed.
+fn remove_bindgen_attrs(attrs: &mut Vec<Attribute>) -> Result<(), ConvertError> {
+    if has_attr(&attrs, "bindgen_unused_template_param") {
+        return Err(ConvertError::UnusedTemplateParam);
+    }
+
     fn is_bindgen_attr(attr: &Attribute) -> bool {
         let segments = &attr.path.segments;
         segments.len() == 1
@@ -37,5 +45,10 @@ fn remove_bindgen_attrs(attrs: &mut Vec<Attribute>) {
                 .starts_with("bindgen_")
     }
 
-    attrs.retain(|a| !is_bindgen_attr(a))
+    attrs.retain(|a| !is_bindgen_attr(a));
+    Ok(())
+}
+
+fn has_attr(attrs: &[Attribute], attr_name: &str) -> bool {
+    attrs.iter().any(|at| at.path.is_ident(attr_name))
 }

--- a/engine/src/conversion/analysis/pod/mod.rs
+++ b/engine/src/conversion/analysis/pod/mod.rs
@@ -25,6 +25,8 @@ use crate::{
         analysis::type_converter::{add_analysis, TypeConversionContext, TypeConverter},
         api::{AnalysisPhase, Api, ApiDetail, TypeKind, TypedefKind, UnanalyzedApi},
         codegen_rs::make_non_pod,
+        convert_error::{ConvertErrorWithContext, ErrorContext},
+        error_reporter::add_api_or_report_error,
         ConvertError,
     },
     types::{Namespace, QualifiedName},
@@ -56,27 +58,33 @@ pub(crate) fn analyze_pod_apis(
     let byvalue_checker = ByValueChecker::new_from_apis(&apis, config)?;
     let mut extra_apis = Vec::new();
     let mut type_converter = TypeConverter::new(config, &apis);
-    let mut results: Vec<_> = apis
-        .into_iter()
-        .map(|api| analyze_pod_api(api, &byvalue_checker, &mut type_converter, &mut extra_apis))
-        .collect::<Result<Vec<_>, ConvertError>>()?;
+    let mut results = Vec::new();
+    for api in apis.into_iter() {
+        let id = api.name().get_final_ident();
+        add_api_or_report_error(api.name(), &mut results, || {
+            analyze_pod_api(api, &byvalue_checker, &mut type_converter, &mut extra_apis)
+                .map_err(|e| ConvertErrorWithContext(e, Some(ErrorContext::Item(id))))
+                .map(Some)
+        })
+    }
     // Conceivably, the process of POD-analysing the first set of APIs could result
     // in us creating new APIs to concretize generic types.
     let mut more_extra_apis = Vec::new();
-    let mut more_results = extra_apis
-        .into_iter()
-        .map(add_analysis)
-        .map(|api| {
+
+    for api in extra_apis.into_iter().map(add_analysis) {
+        let id = api.name().get_final_ident();
+        add_api_or_report_error(api.name(), &mut results, || {
             analyze_pod_api(
                 api,
                 &byvalue_checker,
                 &mut type_converter,
                 &mut more_extra_apis,
             )
+            .map_err(|e| ConvertErrorWithContext(e, Some(ErrorContext::Item(id))))
+            .map(Some)
         })
-        .collect::<Result<Vec<_>, ConvertError>>()?;
+    }
     assert!(more_extra_apis.is_empty());
-    results.append(&mut more_results);
     Ok(results)
 }
 
@@ -108,7 +116,7 @@ fn analyze_pod_api(
             mut item,
             analysis: _,
         } => {
-            super::remove_bindgen_attrs(&mut item.attrs);
+            super::remove_bindgen_attrs(&mut item.attrs)?;
             let analysis = if byvalue_checker.is_pod(&ty_id) {
                 TypeKind::Pod
             } else {
@@ -120,7 +128,7 @@ fn analyze_pod_api(
             mut item,
             analysis: _,
         } => {
-            super::remove_bindgen_attrs(&mut item.attrs);
+            super::remove_bindgen_attrs(&mut item.attrs)?;
             let type_kind = if byvalue_checker.is_pod(&ty_id) {
                 // It's POD so let's mark dependencies on things in its field
                 get_struct_field_types(

--- a/engine/src/conversion/analysis/pod/mod.rs
+++ b/engine/src/conversion/analysis/pod/mod.rs
@@ -25,8 +25,7 @@ use crate::{
         analysis::type_converter::{add_analysis, TypeConversionContext, TypeConverter},
         api::{AnalysisPhase, Api, ApiDetail, TypeKind, TypedefKind, UnanalyzedApi},
         codegen_rs::make_non_pod,
-        convert_error::{ConvertErrorWithContext, ErrorContext},
-        error_reporter::add_api_or_report_error,
+        error_reporter::convert_item_apis,
         ConvertError,
     },
     types::{Namespace, QualifiedName},
@@ -59,31 +58,21 @@ pub(crate) fn analyze_pod_apis(
     let mut extra_apis = Vec::new();
     let mut type_converter = TypeConverter::new(config, &apis);
     let mut results = Vec::new();
-    for api in apis.into_iter() {
-        let id = api.name().get_final_ident();
-        add_api_or_report_error(api.name(), &mut results, || {
-            analyze_pod_api(api, &byvalue_checker, &mut type_converter, &mut extra_apis)
-                .map_err(|e| ConvertErrorWithContext(e, Some(ErrorContext::Item(id))))
-                .map(Some)
-        })
-    }
+    convert_item_apis(apis, &mut results, |api| {
+        analyze_pod_api(api, &byvalue_checker, &mut type_converter, &mut extra_apis).map(Some)
+    });
     // Conceivably, the process of POD-analysing the first set of APIs could result
     // in us creating new APIs to concretize generic types.
     let mut more_extra_apis = Vec::new();
-
-    for api in extra_apis.into_iter().map(add_analysis) {
-        let id = api.name().get_final_ident();
-        add_api_or_report_error(api.name(), &mut results, || {
-            analyze_pod_api(
-                api,
-                &byvalue_checker,
-                &mut type_converter,
-                &mut more_extra_apis,
-            )
-            .map_err(|e| ConvertErrorWithContext(e, Some(ErrorContext::Item(id))))
-            .map(Some)
-        })
-    }
+    convert_item_apis(extra_apis, &mut results, |api| {
+        analyze_pod_api(
+            add_analysis(api),
+            &byvalue_checker,
+            &mut type_converter,
+            &mut more_extra_apis,
+        )
+        .map(Some)
+    });
     assert!(more_extra_apis.is_empty());
     Ok(results)
 }

--- a/engine/src/conversion/analysis/tdef.rs
+++ b/engine/src/conversion/analysis/tdef.rs
@@ -112,7 +112,9 @@ fn get_replacement_typedef(
     deps: &mut HashSet<QualifiedName>,
 ) -> Result<ApiDetail<TypedefAnalysis>, ConvertErrorWithContext> {
     let mut converted_type = ity.clone();
-    remove_bindgen_attrs(&mut converted_type.attrs);
+    let id = ity.ident.clone();
+    remove_bindgen_attrs(&mut converted_type.attrs)
+        .map_err(|e| ConvertErrorWithContext(e, Some(ErrorContext::Item(id))))?;
     let type_conversion_results = type_converter.convert_type(
         (*ity.ty).clone(),
         name.get_namespace(),

--- a/engine/src/conversion/error_reporter.rs
+++ b/engine/src/conversion/error_reporter.rs
@@ -48,23 +48,53 @@ where
 /// Run some code which generates an API. Add that API, or if
 /// anything goes wrong, instead add a note of the problem in our
 /// output API such that users will see documentation for the problem.
-pub(crate) fn add_api_or_report_error<F, A>(tn: QualifiedName, apis: &mut Vec<Api<A>>, fun: F)
+pub(crate) fn convert_apis<F, A, B>(in_apis: Vec<Api<A>>, out_apis: &mut Vec<Api<B>>, mut fun: F)
 where
-    F: FnOnce() -> Result<Option<Api<A>>, ConvertErrorWithContext>,
+    F: FnMut(Api<A>) -> Result<Option<Api<B>>, ConvertErrorWithContext>,
     A: AnalysisPhase,
+    B: AnalysisPhase,
 {
-    match fun() {
-        Ok(Some(api)) => {
-            apis.push(api);
+    out_apis.extend(in_apis.into_iter().filter_map(|api| {
+        let tn = api.name();
+        match fun(api) {
+            Ok(opt) => opt,
+            Err(ConvertErrorWithContext(err, None)) => {
+                eprintln!("Ignored {}: {}", tn.to_string(), err);
+                None
+            }
+            Err(ConvertErrorWithContext(err, Some(ctx))) => {
+                eprintln!("Ignored {}: {}", tn.to_string(), err);
+                Some(ignored_item(tn.get_namespace(), ctx, err))
+            }
         }
-        Ok(None) => {}
-        Err(ConvertErrorWithContext(err, None)) => {
-            eprintln!("Ignored {}: {}", tn.to_string(), err);
-        }
-        Err(ConvertErrorWithContext(err, Some(ctx))) => {
-            eprintln!("Ignored {}: {}", tn.to_string(), err);
-            push_ignored_item(tn.get_namespace(), ctx, err, apis);
-        }
+    }))
+}
+
+/// Run some code which generates an API for an item (as opposed to
+/// a method). Add that API, or if
+/// anything goes wrong, instead add a note of the problem in our
+/// output API such that users will see documentation for the problem.
+pub(crate) fn convert_item_apis<F, A, B>(
+    in_apis: Vec<Api<A>>,
+    out_apis: &mut Vec<Api<B>>,
+    mut fun: F,
+) where
+    F: FnMut(Api<A>) -> Result<Option<Api<B>>, ConvertError>,
+    A: AnalysisPhase,
+    B: AnalysisPhase,
+{
+    convert_apis(in_apis, out_apis, |api| {
+        let id = api.name().get_final_ident();
+        fun(api).map_err(|e| ConvertErrorWithContext(e, Some(ErrorContext::Item(id))))
+    })
+}
+
+fn ignored_item<A: AnalysisPhase>(ns: &Namespace, ctx: ErrorContext, err: ConvertError) -> Api<A> {
+    Api {
+        name: QualifiedName::new(ns, ctx.get_id().clone()),
+        original_name: None,
+        deps: HashSet::new(),
+        detail: ApiDetail::IgnoredItem { err, ctx },
     }
 }
 
@@ -74,10 +104,5 @@ fn push_ignored_item(
     err: ConvertError,
     apis: &mut Vec<Api<impl AnalysisPhase>>,
 ) {
-    apis.push(Api {
-        name: QualifiedName::new(ns, ctx.get_id().clone()),
-        original_name: None,
-        deps: HashSet::new(),
-        detail: ApiDetail::IgnoredItem { err, ctx },
-    });
+    apis.push(ignored_item(ns, ctx, err));
 }

--- a/engine/src/integration_tests.rs
+++ b/engine/src/integration_tests.rs
@@ -5160,23 +5160,14 @@ fn test_mbstate() {
 }
 
 #[test]
-#[ignore] // https://github.com/google/autocxx/issues/470
-fn test_issue_470() {
+fn test_issue_470_492() {
     let hdr = indoc! {"
         namespace std {
-        template <bool, typename _Iftrue, typename _Iffalse> struct a;
+            template <bool, typename _Iftrue, typename _Iffalse> struct a;
         }
-        template <typename> struct b { template <typename> struct c; };
-        struct d : b<d> {};
-        struct e : b<e> {};
-        struct f : b<f> {};
-        struct g : b<g> {};
-        struct h;
-        template <typename... i> struct j {
-          typedef std::a<d::c<i...>::k, d,
-                         std::a<e::c<i...>::k, e,
-                                std::a<f::c<i...>::k, f, std::a<g::c<i...>::k, g, h>>>>
-              l;
+        template <typename> struct b;
+        template <typename d> struct c {
+            typedef std::a<b<d>::c, int, int> e;
         };
     "};
     let rs = quote! {};

